staging/vhost-device-console/src/vhu_console.rs

```diff
@@ -0,0 +1,1412 @@
+    #[error("Invalid descriptor count {0}")]
+    UnexpectedDescriptorCount(usize),
+    #[error("Invalid descriptor size, expected: {0}, found: {1}")]
+    UnexpectedDescriptorSize(usize, u32),
+    #[error("Descriptor not found")]
+    DescriptorNotFound,
+    #[error("Failed to send notification")]
+    NotificationFailed,
+    #[error("Descriptor read failed")]
+    DescriptorReadFailed,
+    #[error("Descriptor write failed")]
+    DescriptorWriteFailed,
+    #[error("Failed to create new EventFd")]
+    EventFdFailed,
+    #[error("Failed to add control message in the internal queue")]
+    RxCtrlQueueAddFailed,
+}
+
+impl convert::From<Error> for io::Error {
+    fn from(e: Error) -> Self {
+        io::Error::new(io::ErrorKind::Other, e)
+    }
+}
+
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for VirtioConsoleControl {}
+
+pub(crate) struct VhostUserConsoleBackend {
+    controller: Arc<RwLock<ConsoleController>>,
+    acked_features: u64,
+    event_idx: bool,
+    rx_ctrl_fifo: Queue<VirtioConsoleControl>,
+    rx_data_fifo: Queue<String>,
+    pub(crate) ready: bool,
+    pub(crate) ready_to_write: bool,
+    pub(crate) output_buffer: String,
+    pub(crate) output_flag: bool,
+    pub(crate) rx_event: EventFd,
+    pub(crate) rx_ctrl_event: EventFd,
+    pub(crate) exit_event: EventFd,
+    mem: Option<GuestMemoryAtomic<GuestMemoryMmap>>,
+}
+
+type ConsoleDescriptorChain = DescriptorChain<GuestMemoryLoadGuard<GuestMemoryMmap<()>>>;
+
+impl VhostUserConsoleBackend {
+    pub(crate) fn new(controller: Arc<RwLock<ConsoleController>>) -> Result<Self> {
+        Ok(VhostUserConsoleBackend {
+            controller,
+            event_idx: false,
+            rx_ctrl_fifo: Queue::new(),
+            rx_data_fifo: Queue::new(),
+            acked_features: 0x0,
+            ready: false,
+            ready_to_write: false,
+            output_buffer: String::new(),
+            output_flag: false,
+            rx_event: EventFd::new(EFD_NONBLOCK).map_err(|_| Error::EventFdFailed)?,
+            rx_ctrl_event: EventFd::new(EFD_NONBLOCK).map_err(|_| Error::EventFdFailed)?,
+            exit_event: EventFd::new(EFD_NONBLOCK).map_err(|_| Error::EventFdFailed)?,
+            mem: None,
+        })
+    }
+
+    fn print_console_frame(&self, control_msg: VirtioConsoleControl) {
+        trace!("id 0x{:x}", control_msg.id.to_native());
+        trace!("event 0x{:x}", control_msg.event.to_native());
+        trace!("value 0x{:x}", control_msg.value.to_native());
+    }
+
+    fn process_rx_requests(
+        &mut self,
+        requests: Vec<ConsoleDescriptorChain>,
+        vring: &VringRwLock,
+    ) -> Result<bool> {
+        trace!("process_rx_requests");
+
+        if requests.is_empty() {
+            trace!("requests.is_empty");
+            return Ok(true);
+        }
+
+        let desc_chain = &requests[0];
```

Comment:
AFAIU, request is an array of `ConsoleDescriptorChain`. This line means that for reception we are only using the chain at 0, right? why?

Reply:
Hello @MatiasVara, in oldest implementations I couldn't make it work otherwise and this code was left like this since then.
The latest version goes as usual through all the elements into the request array.